src/tools/rbd_mirror/InstanceReplayer.cc

```diff
@@ -0,0 +1,494 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "include/stringify.h"
+#include "common/Timer.h"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "librbd/Utils.h"
+#include "ImageReplayer.h"
+#include "InstanceReplayer.h"
+#include "Threads.h"
+
+#define dout_context g_ceph_context
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd::mirror::InstanceReplayer: " \
+                           << this << " " << __func__ << ": "
+
+namespace rbd {
+namespace mirror {
+
+using librbd::util::create_async_context_callback;
+using librbd::util::create_context_callback;
+
+template <typename I>
+InstanceReplayer<I>::InstanceReplayer(
+  Threads<I> *threads, std::shared_ptr<ImageDeleter> image_deleter,
+  ImageSyncThrottlerRef<I> image_sync_throttler, RadosRef local_rados,
+  const std::string &local_mirror_uuid, int64_t local_pool_id)
+  : m_threads(threads), m_image_deleter(image_deleter),
+    m_image_sync_throttler(image_sync_throttler), m_local_rados(local_rados),
+    m_local_mirror_uuid(local_mirror_uuid), m_local_pool_id(local_pool_id),
+    m_lock("rbd::mirror::InstanceReplayer " + stringify(local_pool_id)) {
+}
+
+template <typename I>
+InstanceReplayer<I>::~InstanceReplayer() {
+  assert(m_image_state_check_task == nullptr);
+  assert(m_async_op_tracker.empty());
+  assert(m_image_replayers.empty());
+}
+
+template <typename I>
+int InstanceReplayer<I>::init() {
+  C_SaferCond init_ctx;
+  init(&init_ctx);
+  return init_ctx.wait();
+}
+
+template <typename I>
+void InstanceReplayer<I>::init(Context *on_finish) {
+  dout(20) << dendl;
+
+  Context *ctx = new FunctionContext(
+    [this, on_finish] (int r) {
+      {
+        Mutex::Locker timer_locker(m_threads->timer_lock);
+        schedule_image_state_check_task();
+      }
+      on_finish->complete(0);
+    });
+
+  m_threads->work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void InstanceReplayer<I>::shut_down() {
+  C_SaferCond shut_down_ctx;
+  shut_down(&shut_down_ctx);
+  int r = shut_down_ctx.wait();
+  assert(r == 0);
+}
+
+template <typename I>
+void InstanceReplayer<I>::shut_down(Context *on_finish) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  assert(m_on_shut_down == nullptr);
+  m_on_shut_down = on_finish;
+
+  Context *ctx = new FunctionContext(
+    [this] (int r) {
+      cancel_image_state_check_task();
+      Mutex::Locker locker(m_lock);
```

Comment:
Actually, the lock is not needed in wait_for_ops. I just wanted to make it similar to other xyz()/handle_xyz() functions.

Ok, I removed the lock in wait_for_ops (and create_async_context_callback, which is not needed after this).